planner/core/rule_partition_processor.go

```diff
@@ -224,73 +224,251 @@ func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.Part
+	colIDToUniqueID map[int64]int64
+	fullRange       map[int]struct{}
+	listPrune       *tables.ForListPruning
+}
 
+func newListPartitionPruner(ctx sessionctx.Context, tbl table.Table, partitionNames []model.CIStr,
+	s *partitionProcessor, conds []expression.Expression, pruneList *tables.ForListPruning) *listPartitionPruner {
 	colIDToUniqueID := make(map[int64]int64)
 	for _, cond := range conds {
 		condCols := expression.ExtractColumns(cond)
 		for _, c := range condCols {
 			colIDToUniqueID[c.ID] = c.UniqueID
 		}
 	}
+	fullRange := make(map[int]struct{})
+	fullRange[FullRange] = struct{}{}
+	return &listPartitionPruner{
+		partitionProcessor: s,
+		ctx:                ctx,
+		pi:                 tbl.Meta().Partition,
+		partitionNames:     partitionNames,
+		colIDToUniqueID:    colIDToUniqueID,
+		fullRange:          fullRange,
+		listPrune:          pruneList,
+	}
+}
 
-	pruneList := partExpr.ForListPruning
-	cols := make([]*expression.Column, 0, len(pruneList.ExprCols))
-	colLen := make([]int, 0, len(pruneList.ExprCols))
-	for _, c := range pruneList.ExprCols {
+func (l *listPartitionPruner) locatePartition(cond expression.Expression) (tables.ListPartitionLocation, bool, error) {
+	switch sf := cond.(type) {
+	case *expression.Constant:
+		b, err := sf.Value.ToBool(l.ctx.GetSessionVars().StmtCtx)
+		if err == nil && b == 0 {
+			// A constant false expression.
+			return nil, false, nil
+		}
+	case *expression.ScalarFunction:
+		switch sf.FuncName.L {
+		case ast.LogicOr:
+			dnfItems := expression.FlattenDNFConditions(sf)
+			return l.locatePartitionByDNFCondition(dnfItems)
+		case ast.LogicAnd:
+			cnfItems := expression.FlattenCNFConditions(sf)
+			return l.locatePartitionByCNFCondition(cnfItems)
+		}
+		return l.locatePartitionByColumn(sf)
+	}
+	return nil, true, nil
+}
+
+func (l *listPartitionPruner) locatePartitionByCNFCondition(conds []expression.Expression) (tables.ListPartitionLocation, bool, error) {
+	if len(conds) == 0 {
+		return nil, true, nil
+	}
+	countFull := 0
+	intersect := tables.NewListPartitionLocationIntersect()
+	for _, cond := range conds {
+		cnfLoc, isFull, err := l.locatePartition(cond)
+		if err != nil {
+			return nil, false, err
+		}
+		if isFull {
+			countFull++
+			continue
+		}
+		if cnfLoc.IsEmpty() {
+			// No partition for intersection, just return 0 partition.
+			return nil, false, nil
+		}
+		if !intersect.Intersect(cnfLoc) {
+			return nil, false, nil
+		}
+	}
+	if countFull == len(conds) {
+		return nil, true, nil
+	}
+	return intersect.GetLocation(), false, nil
+}
+
+func (l *listPartitionPruner) locatePartitionByDNFCondition(conds []expression.Expression) (tables.ListPartitionLocation, bool, error) {
+	if len(conds) == 0 {
+		return nil, true, nil
+	}
+	merger := tables.NewListPartitionLocationMerger()
+	for _, cond := range conds {
+		dnfLoc, isFull, err := l.locatePartition(cond)
+		if err != nil || isFull {
+			return nil, isFull, err
+		}
+		merger.Merge(dnfLoc)
```

Comment:
It's better to name it `Union`, because `intersect` and `union` are common set operations, `Merge` looks weird.

Reply:
great catch, done.